pkg/health/membergarbagecollector/membergarbagecollector.go

```diff
@@ -0,0 +1,179 @@
+		return nil, &errors.EtcdError{
+			Message: "nil kubernetes clientset passed, can not create heartbeat",
+		}
+	}
+	namespace, err := utils.GetEnvVarOrError(podNamespace)
+	if err != nil {
+		logger.Fatalf("POD_NAMESPACE env var not present: %v", err)
+	}
+	podName, err := utils.GetEnvVarOrError(podName)
+	if err != nil {
+		logger.Fatalf("POD_NAME env var not present: %v", err)
+	}
+	return &MemberGarbageCollector{
+		logger:       logger.WithField("actor", "membergc"),
+		k8sClient:    clientSet,
+		podNamespace: namespace,
+		etcdName:     podName[:strings.LastIndex(podName, "-")], //TODO Evaluate if this will always be sufficient
+	}, nil
+}
+
+// RunMemberGarbageCollectorPeriodically periodically calls the member garbage collector
+func RunMemberGarbageCollectorPeriodically(ctx context.Context, stopCh chan struct{}, hconfig *brtypes.HealthConfig, logger *logrus.Entry, etcdConfig *brtypes.EtcdConnectionConfig) {
+	clientSet, err := miscellaneous.GetKubernetesClientSetOrError()
+	if err != nil {
+		logger.Errorf("failed to create kubernetes clientset: %v", err)
+		return
+	}
+	mgc, err := NewMemberGarbageCollector(logger, clientSet)
+	if err != nil {
+		logger.Errorf("failed to initialize new heartbeat: %v", err)
+		return
+	}
+	mgc.gcTimer = time.NewTimer(hconfig.MemberGCDuration.Duration)
+	defer mgc.gcTimer.Stop()
+	mgc.logger.Info("Started etcd member garbage collector")
+
+	for {
+		//Create etcd client to get etcd cluster member list
+		clientFactory := etcdutil.NewFactory(*etcdConfig)
+		etcdCluster, err := clientFactory.NewCluster()
+		if err != nil {
+			mgc.logger.Errorf("Failed to create etcd cluster client: %v", err)
+		}
+		defer etcdCluster.Close()
+		select {
+		case <-mgc.gcTimer.C:
+			err := mgc.RemoveSuperfluousMembers(ctx, etcdCluster)
+			if err != nil {
+				mgc.logger.Warn(err)
+			}
+			mgc.gcTimer.Reset(hconfig.MemberGCDuration.Duration)
+		case <-ctx.Done():
+			mgc.logger.Info("Stopping etcd member garbage collector")
+			etcdCluster.Close()
+			return
+		case <-stopCh:
+			mgc.logger.Info("Stopping etcd member garbage collector")
+			return
+		}
+	}
+}
+
+// RemoveSuperfluousMembers removes a member from the etcd cluster if its corresponding pod is not present
+func (mgc *MemberGarbageCollector) RemoveSuperfluousMembers(ctx context.Context, etcdCluster etcdclient.ClusterCloser) error {
+	etcd := &druidv1alpha1.Etcd{}
+	if err := mgc.k8sClient.Get(ctx, client.ObjectKey{
+		Namespace: mgc.podNamespace,
+		Name:      mgc.etcdName,
+	}, etcd); err != nil {
+		return &errors.EtcdError{
+			Message: fmt.Sprintf("Could not fetch etcd %v with error: %v", mgc.etcdName, err),
+		}
+	}
+	// Return if no etcd member present
+	if len(etcd.Status.Members) < 1 {
+		return &errors.EtcdError{
+			Message: "There are no members present in the etcd status. Nothing to clean",
+		}
+	}
+
+	// Return if memberlist size <= desired cluster size. Cleanup not needed
+	if len(etcd.Status.Members) <= etcd.Spec.Replicas {
+		return nil
+	}
```

Comment:
We can't return here since due to a race the `Members` slice can already be updated by the Druid, i.e. `etcd.Status.Members == etcd.Spec.Replicas` but the members themselves are still in the cluster.

Reply:
Thanks for catching this 👍🏻 
Removed unneeded condition